src/target/at32f43x.c

```diff
@@ -0,0 +1,376 @@
+	if ((target->cpuid & CPUID_PARTNO_MASK) != CORTEX_M4)
+		return false;
+
+	// Artery chips use the complete idcode word for identification
+	const uint32_t idcode = target_mem_read32(target, DBGMCU_IDCODE);
+	const uint32_t series = idcode & AT32F4x_IDCODE_SERIES_MASK;
+	const uint16_t part_id = idcode & AT32F4x_IDCODE_PART_MASK;
+
+	if (series == AT32F43_SERIES_4K || series == AT32F43_SERIES_2K)
+		return at32f43_detect(target, part_id);
+	return false;
+}
+
+static bool at32f43_flash_unlock(target_s *const target, const uint32_t bank_reg_offset)
+{
+	if (target_mem_read32(target, AT32F435_FLASH_CTRL + bank_reg_offset) & FLASH_CTRL_OPLK) {
+		/* Enable FLASH operations in requested bank */
+		target_mem_write32(target, AT32F435_FLASH_UNLOCK + bank_reg_offset, KEY1);
+		target_mem_write32(target, AT32F435_FLASH_UNLOCK + bank_reg_offset, KEY2);
+	}
+	const uint32_t ctrlx = target_mem_read32(target, AT32F435_FLASH_CTRL + bank_reg_offset);
+	if (ctrlx & FLASH_CTRL_OPLK)
+		DEBUG_ERROR("%s failed, CTRLx: 0x%08" PRIx32 "\n", __func__, ctrlx);
+	return !(ctrlx & FLASH_CTRL_OPLK);
+}
+
+static inline void at32f43_flash_clear_eop(target_s *const target, const uint32_t bank_reg_offset)
+{
+	const uint32_t status = target_mem_read32(target, AT32F435_FLASH_STS + bank_reg_offset);
+	target_mem_write32(target, AT32F435_FLASH_STS + bank_reg_offset, status | FLASH_STS_ODF); /* ODF is W1C */
+}
+
+static bool at32f43_flash_busy_wait(
+	target_s *const target, const uint32_t bank_reg_offset, platform_timeout_s *const timeout)
+{
+	/* Read FLASH_STS to poll for Operation Busy Flag */
+	uint32_t status = FLASH_STS_OBF;
+	/* Checking for ODF/EOP requires methodically clearing the ODF */
+	while (!(status & FLASH_STS_ODF) && (status & FLASH_STS_OBF)) {
+		status = target_mem_read32(target, AT32F435_FLASH_STS + bank_reg_offset);
+		if (target_check_error(target)) {
+			DEBUG_ERROR("Lost communications with target\n");
+			return false;
+		}
+		if (timeout)
+			target_print_progress(timeout);
+	}
+	if (status & FLASH_STS_PRGMERR) {
+		DEBUG_ERROR("at32f43 flash error, STS: 0x%" PRIx32 "\n", status);
+		return false;
+	}
+	return !(status & FLASH_STS_PRGMERR);
+}
+
+static inline uint32_t at32f43_bank_offset_for(const target_addr_t addr, const target_addr_t bank_split)
+{
+	if (addr >= bank_split)
+		return FLASH_BANK2_REG_OFFSET;
+	return FLASH_BANK1_REG_OFFSET;
+}
+
+static bool at32f43_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len)
+{
+	target_s *target = flash->t;
+	const at32f43_flash_s *const sf = (at32f43_flash_s *)flash;
+
+	/* Erase range begins in bank 1? Unlock bank 1 */
+	if (addr < sf->bank_split && !at32f43_flash_unlock(target, 0))
+		return false;
+	/* Erase range ends in bank 2? Unlock bank 2 */
+	const target_addr_t end = addr + len - 1U;
+	if (end >= sf->bank_split && !at32f43_flash_unlock(target, FLASH_BANK2_REG_OFFSET))
+		return false;
+
+	for (size_t offset = 0; offset < len; offset += flash->blocksize) {
```

Comment:
The target Flash API layer guarantees `len` will be flash->blocksize and `addr` will start on a block boundary, so this loop can be ditched and the return simplified to `return at32f43_flash_busy_wait(target, bank_reg_offset, NULL);`

Reply:
This loop was copied verbatim from `stm32f1_flash_erase()` by me not familiar with internal BMD API layers. Do you mean that in current implementation this for-loop will always run exactly once and its contents can be de-indented along with dropping `offset`? I may need to test with `DEBUG_INFO`s before significantly changing driver logic.

Reply:
Exactly, and the STM32F1 implementation hasn't yet been updated to match that reality.

Reply:
co-signing this note. the `stm32f1_flash_erase` is "legacy code" that needs to be updated to the current "workflow"

Testing is very much welcome as we can't have too many guarantees that code is working properly, but it is expected to function as noted. (I believe there is come testing "logs" on a discord thread when I originally wrote that code)